File: knuth_morris_pratt_kmp.py

def preprocess(pattern):
    kmp_arr = [0]*len(pattern)
    j = 0
    for i in range(1, len(pattern)):
        while True:
            if pattern[j] == pattern[i]:
                kmp_arr[i] = j + 1
                j += 1
                break
            else:
                if j == 0:
                    kmp_arr[i] = j
                    break
                else:
                    j = kmp_arr[j-1]
    return kmp_arr


def match(text, pattern, kmp_arr):
    match_indices = []
    p = 0
    for t in range(0, len(text)):
        while True:
            if text[t] == pattern[p]:
                if p == (len(pattern) - 1):
                    match_indices.append(t-p)
                    p = kmp_arr[p]
                else:
                    p += 1
                break
            else:
                if p == 0:
                    break
                else:
                    p = kmp_arr[p-1]
    return match_indices


def main():
    text = "abcxabcdabxabcdabcdabcyabcabcdabcyy"
    pattern = "abcdabcy"
    kmp_array = preprocess(pattern)
    print('kmp preprocessed array for ', pattern, ':', kmp_array)
    matches = match(text, pattern, kmp_array)
    print('matched indices : ', matches)


if (__name__ == "__main__"):
    main()

File: cpp/knuth_morris_pratt_kmp.cc
#include <iostream>
#include <string>
#include <vector>

#define DEBUG

std::vector<int> preprocess(std::string pattern) {
  std::vector<int> kmp_arr(pattern.size());
  int j = 0;
  int i = 1;
  kmp_arr[0] = 0;
  for (i = 1; i < pattern.size(); i++) {
    while (true) {
      if (pattern[j] == pattern[i]) {
        kmp_arr[i] = j + 1;
        j++;
        break;
      } else {
        if (j == 0) {
          kmp_arr[i] = 0;
          break;
        }
        j = kmp_arr[j - 1];
      }
    }
  }
  return kmp_arr;
}

std::vector<int> match(std::string text,
                       std::string pattern,
                       std::vector<int> kmp_arr) {
  if (kmp_arr.size() != pattern.size()) {
    std::cout << "ERROR: KMP array size doesn't match pattern string size"
              << std::endl;
    return std::vector<int>();
  }

  std::vector<int> res;
  int p = 0;
  for (int t = 0; t < text.size(); t++) {
    while (true) {
      if (text[t] == pattern[p]) {
        if (p == pattern.size() - 1) {
          res.push_back(t - p);
          p = kmp_arr[p];
        } else {
          p++;
        }
        break;
      } else {
        if (p == 0) {
          break;
        } else {
          p = kmp_arr[p - 1];
        }
      }
    }
  }
  return res;
}

int main() {
  std::string text = "abcxabcdabxabcdabcdabcyabcabcdabcyy";
  std::string pattern = "abcdabcy";
  std::vector<int> kmp_array = preprocess(pattern);

#ifdef DEBUG
  std::cout << "KMP preprocessed array for pattern(" << pattern << ") : ";
  for (int i = 0; i < kmp_array.size(); i++) {
    std::cout << kmp_array[i] << " ";
  }
  std::cout << std::endl;
#endif

  std::cout << "match positions : ";
  std::vector<int> match_indexes = match(text, pattern, kmp_array);
  for (int i = 0; i < match_indexes.size(); i++) {
    std::cout << match_indexes[i] << " ";
  }
  std::cout << std::endl;
}
